src/databricks/labs/ucx/source_code/base.py

```diff
@@ -170,3 +181,41 @@ def __init__(self, linters: list[Linter]):
     def lint(self, code: str) -> Iterable[Advice]:
         for linter in self._linters:
             yield from linter.lint(code)
+
+
+class PythonSequentialLinter(Linter):
+
+    def __init__(self, linters: list[PythonLinter]):
+        self._linters = linters
+        self._tree: Tree | None = None
+
+    def lint(self, code: str) -> Iterable[Advice]:
+        try:
+            tree = self._parse_and_append(code)
+            yield from self.lint_tree(tree)
+        except AstroidSyntaxError as e:
+            yield Failure('syntax-error', str(e), 0, 0, 0, 0)
+
+    def lint_tree(self, tree: Tree) -> Iterable[Advice]:
+        for linter in self._linters:
+            yield from linter.lint_tree(tree)
+
+    def _parse_and_append(self, code: str) -> Tree:
+        tree = Tree.normalize_and_parse(code)
+        self.append_tree(tree)
+        return tree
+
+    def append_tree(self, tree):
```

Comment:
why do we need this here? why can't we simply process a collection of cells into a `Tree` and work with that? it's a more straightforward implementation that is easy to read

Reply:
The rationale is that we need to lint on a per cell basis, so we need to remember the tree for each cell

Reply:
The 'full tree' includes child notebooks, and we definitely don't want to lint these as part of linting a notebook. We only need the code for value inference.